### DISCUSSION TITLE
How to subscribe to blocks and decode the extrinsics?

### DISCUSSION THREAD

**Discussion by f-squirrel** [smol-dot/smoldot#817] [All]
I am trying to use smoldot lightweight client in Rust to subscribe to the latest updates in the chain (Polkadot).
I use the example to subscribe to blockheads but receive them as strings. Is there any crate with Polkadot primitives that I can use to deserialize extrinsics to actual Polkadot data structures?

**Comment by tomaka** [comment]
I don't think there's any Rust crate apart from Substrate itself that lets you decode extrinsics.

In order to do that, the preliminary step is to query and decode the so-called metadata, as it contains information about how to decode extrinsics.

**Reply by f-squirrel** [reply]
@tomaka, thank you for getting back to me! I need to understand how I should do the following.
I have a list of addresses/accounts which I need to follow. I need to parse each block and check whether it affects any of my addresses. I really do not want to parse JSON data myself; it is super bug-prone.
In order to do it, I need to be able to parse the blocks. I know there is a way to subscribe to updates of addresses, but I am not sure if this works if there are thousands of addresses.

Am I the first one to implement such a feature for Polkadot?

**Reply by tomaka** [reply]
> Am I the first one to implement such a feature for Polkadot?

You might be the first one to implement this *in Rust*.
To do this, you'd typically use a JavaScript library like PolkadotJS.

**Comment by f-squirrel** [comment]
> > Am I the first one to implement such a feature for Polkadot?
> 
> You might be the first one to implement this _in Rust_. To do this, you'd typically use a JavaScript library like PolkadotJS.

This is very sad:)
I am not familiar with Substrate. Do you know what parts/subcrates I should look at?

**Reply by tomaka** [reply]
After some googling, you might be interested in https://github.com/scs/substrate-api-client
It doesn't support smoldot, but you could start a WebSocket server on top of smoldot and connect to the API to it.
